use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py

```diff
@@ -0,0 +1,264 @@
+"""
+PyTorch implementation of NSGA-II.
+"""
+
+import random
+import shutil
+from pathlib import Path
+
+from tqdm import tqdm
+import numpy as np
+import pandas as pd
+import torch
+
+from data import constants
+from data.eluc_data import ELUCEncoder
+from data.torch_data import TorchDataset
+from predictors.predictor import Predictor
+from prescriptors.prescriptor import Prescriptor
+from prescriptors.nsga2 import nsga2_utils
+from prescriptors.nsga2.candidate import Candidate
+
+class TorchPrescriptor(Prescriptor):
+    """
+    Handles prescriptor candidate evolution
+    """
+    def __init__(self,
+                 pop_size: int,
+                 n_generations: int,
+                 p_mutation: float,
+                 eval_df: pd.DataFrame,
+                 encoder: ELUCEncoder,
+                 predictor: Predictor,
+                 batch_size: int,
+                 candidate_params: dict,
+                 seed_dir=None):
+        
+        self.candidate_params = candidate_params
+        self.pop_size = pop_size
+        self.n_generations = n_generations
+        self.p_mutation = p_mutation
+        self.seed_dir=seed_dir
+
+        self.eval_df = eval_df
+        self.encoded_eval_df = encoder.encode_as_df(eval_df)
+        self.encoder = encoder
+        context_ds = TorchDataset(self.encoded_eval_df[constants.CAO_MAPPING["context"]].to_numpy(),
+                                  np.zeros((len(self.encoded_eval_df), len(constants.RECO_COLS))))
+        self.context_dl = torch.utils.data.DataLoader(context_ds, batch_size=batch_size, shuffle=False)
```

Comment:
We cache the evaluation dataset so that we don't have to re-convert from pandas to torch over and over during training. Then during inference we can pass our own dataframe in.

Reply:
Worth a comment in the code directly.